instance/tests/models/test_instance.py

```diff
@@ -1027,3 +1084,112 @@ def deploy():
         self.assertEqual(mocks.mock_provision_mysql.call_count, 1)
         self.assertEqual(mocks.mock_provision_mongo.call_count, 1)
         self.assertEqual(mocks.mock_provision_swift.call_count, 1)
+
+
+@ddt
+class InstanceStatusTestCase(TestCase):
+    """
+    Test cases for status switching in instance models
+    """
+
+    def _assert_status_conditions(self, instance, is_steady_state=True, is_healthy_state=True):
+        """
+        Assert that status conditions for instance hold as specified
+        """
+        self.assertEqual(instance.status.is_steady_state, is_steady_state)
+        self.assertEqual(instance.status.is_healthy_state, is_healthy_state)
+
+    def test_status_transitions(self):
+        """
+        Test that status transitions work as expected for different instance workflows
+        """
+        # Normal workflow
+        instance = SingleVMOpenEdXInstanceFactory()
+        self.assertEqual(instance.status, Instance.Status.New)
+        self._assert_status_conditions(instance)
+
+        instance._status_to_waiting_for_server()
+        self.assertEqual(instance.status, Instance.Status.WaitingForServer)
+        self._assert_status_conditions(instance, is_steady_state=False)
+
+        instance._status_to_configuring_server()
+        self.assertEqual(instance.status, Instance.Status.ConfiguringServer)
+        self._assert_status_conditions(instance, is_steady_state=False)
+
+        instance._status_to_running()
+        self.assertEqual(instance.status, Instance.Status.Running)
+        self._assert_status_conditions(instance)
+
+        instance._status_to_terminated()
+        self.assertEqual(instance.status, Instance.Status.Terminated)
+        self._assert_status_conditions(instance)
+
+        # Server creation fails
+        instance_bad_server = WaitingSingleVMOpenEdXInstanceFactory()
+        instance_bad_server._status_to_error()
+        self.assertEqual(instance_bad_server.status, Instance.Status.Error)
+        self._assert_status_conditions(instance_bad_server, is_healthy_state=False)
+
+        # Provisioning fails
+        instance_provisioning_failed = ConfiguringSingleVMOpenEdXInstanceFactory()
+        instance_provisioning_failed._status_to_configuration_failed()
+        self.assertEqual(instance_provisioning_failed.status, Instance.Status.ConfigurationFailed)
+        self._assert_status_conditions(instance_provisioning_failed, is_healthy_state=False)
+
+        # Reprovisioning from steady states
```

Comment:
@bradenmacdonald FYI, I'm not using `ddt` here because I wanted to be able to cover all happy paths (workflows) in a single test, without having portions of the test that don't depend on changing data run more than once.